lib/src/main/scala/spinal/lib/eda/fusesoc/FuseSocGeneratorBuilder.scala

```diff
@@ -0,0 +1,126 @@
+    output: GeneratorOutput,
+    spinal_parameter: P
+)
+
+abstract class FusesocRunner[C <: Component, P: ClassTag] {
+  private val yamlMapper = new ObjectMapper(new YAMLFactory().disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
+    .registerModule(DefaultScalaModule) :: ClassTagExtensions
+
+  def run(args: Array[String]): Unit = {
+    val corePath = parseCli(args)
+    val file = new java.io.FileReader(corePath)
+
+    val p = parseCore(file)
+    val sourceOutput =
+      p.output.files.filter(_.head._2.exists(attr => attr._1 == "file_type" && FusesocUtil.checkAvailFileType(attr._2)))
+    assert(sourceOutput.size == 1, "only support 1 output source file")
+    assert(p.target_directory.startsWith("/"), s"target_directory: ${p.target_directory} why is not absolute?")
+
+    val targetDirectory = p.target_directory
+    val mode = sourceOutput.head.head._2.head._2 match {
+      case "verilogSource"       => spinal.core.Verilog
+      case "systemVerilogSource" => spinal.core.SystemVerilog
+      case "vhdlSource"          => spinal.core.VHDL
+      case _ => throw new Exception(s"file_type: ${p.output.files.head.head._2.head._2} not supported filetype")
+    }
+    val netlistFileName = sourceOutput.head.head._1
+
+    val config = SpinalConfig(mode = mode, targetDirectory = targetDirectory, netlistFileName = netlistFileName)
+    config.generate(buildComponent(p.spinal_parameter))
+  }
+
+  def buildComponent(parameter: P): C
+
+  private def parseCore(reader: java.io.Reader): GeneratorParameters[P] =
+    yamlMapper.readValue[GeneratorParameters[P]](reader)
+
+  private def parseCli(args: Array[String]): String = {
+    val builder = scopt.OParser.builder[String]
+    val parser = scopt.OParser.sequence(
+      builder.opt[String]("core_file_path").required().action((x, _) => x)
+    )
+    scopt.OParser.parse(parser, args, null).getOrElse(throw new Exception(""))
+  }
+
+}
+
+class GeneratorBuilder[C <: Component, P: ClassTag](
+    componentName: String,
+    defaultPram: P,
+    descriptions: String = "",
+    version: String = "0.0.0"
+) {
+  private val yamlMapper = new ObjectMapper(new YAMLFactory().disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
+  yamlMapper.registerModule(DefaultScalaModule)
+
+  def addParameters(k: String, v: String): this.type = {
+    output.parameters.update(k, v)
+    this
+  }
+
+  def addFiles(file_name: String, file_type: String): this.type = {
+    output.files.append(new Map1(file_name, Map("file_type" -> file_type)))
+    this
+  }
+  private val output = new GeneratorOutput
+
+  // default add a source file(which will be generated)
+  addFiles(componentName + ".v", "verilogSource")
+
+  private val CoreFile = new Object {
+    val name: String = s"::$componentName:$version"
+    val description: String = descriptions
+    val filesets = new Object {
+      val base: Map[String, List[String]] = Map("depend" -> List("chenbosoft:utils:generators:0.0.0"))
```

Comment:
Is this necessary? what does version 0.0.0 refer to? How about release a start version at 0.0.1?